### PR TITLE
[local_auth] Fix failed biometric authentication not throwing error

### PR DESCRIPTION
**Pull request by LouiseHsu**
On iPhones which support TouchID, when failing biometric authentication multiple times, no error will be thrown. This is because the error code that is being thrown is not being handled - this PR fixes that issues, and also ensures that all error codes regardless if known will be handled and thus throw an appropriate error. This also touches old tests, which were using dummy errors codes - they have now been replaced with appropriate errors for the test.

*List which issues are fixed by this PR. You must list at least one issue.*

Fixes https://github.com/flutter/flutter/issues/103380

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 1]
```diff
@@ -41,6 +41,10 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
     [self deviceSupportsBiometrics:result];
   } else if ([@"isDeviceSupported" isEqualToString:call.method]) {
     result(@YES);
+  } else if ([@"handleAuthReplyWithSuccess" isEqualToString:call.method]) {
+    bool success = [call.arguments[@"success"] boolValue];
```
Comment (hellohuanlin) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:45]: BOOL

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 2]
```diff
@@ -41,6 +41,10 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
...
+    NSError* error = call.arguments[@"error"];
```
Comment (hellohuanlin) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:46]: is `call.arguments[@"error"]` indeed `NSError` type?

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 3]
```diff
@@ -216,41 +220,35 @@ - (void)handleAuthReplyWithSuccess:(BOOL)success
...
+            self->_lastResult = result;
+        } else {
+            result(@NO);
+        }
+        return;
       case LAErrorPasscodeNotSet:
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"
-      // TODO(stuartmorgan): Remove the pragma and s/TouchID/Biometry/ in these constants when
-      // iOS 10 support is dropped. The values are the same, only the names have changed.
```
Comment (hellohuanlin) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: is ios 10 dropped?

**Comment by jmagman** [comment]
Removing myself until this out of draft and ready to review 🙂

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 4]
```diff
@@ -216,41 +220,35 @@ - (void)handleAuthReplyWithSuccess:(BOOL)success
     result(@YES);
   } else {
     switch (error.code) {
+      case LAErrorSystemCancel:
+        if ([arguments[@"stickyAuth"] boolValue]) {
+            self->_lastCallArgs = arguments;
+            self->_lastResult = result;
```
Comment (hellohuanlin) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:226]: Do we need `result(@YES)` here?

**Comment by jmagman** [comment]
> List which issues are fixed by this PR. You must list at least one issue.
> 
> https://github.com/flutter/flutter/issues/103380

Change this to "fixes: https://github.com/flutter/flutter/issues/103380"  to link the issue for auto-close 
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 5]
```diff
@@ -216,41 +220,35 @@ - (void)handleAuthReplyWithSuccess:(BOOL)success
...
-      case LAErrorTouchIDNotAvailable:
-      case LAErrorTouchIDNotEnrolled:
-      case LAErrorTouchIDLockout:
-#pragma clang diagnostic pop
+      case LAErrorAuthenticationFailed:
+      case LAErrorBiometryNotAvailable:
+      case LAErrorBiometryNotEnrolled:
+      case LAErrorBiometryLockout:
       case LAErrorUserFallback:
+      default:
```
Comment (hellohuanlin) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:237]: when will this default be called?

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 3 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: Yes (?), I was under the impression minimum ios version is 11.0 now

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 5 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:237]: So I added the default so it would catch any unknown error codes and throw an error - the original issue was happening because an error code was thrown that wasnt listed and it was failing silently.  Should I change my approach?

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 4 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:226]: I think result(@Yes) should only get returned if it's successful, so if the system cancels it should return no error and exit silently

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 2 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:46]: It should be!
https://github.com/flutter/plugins/blob/abb8e681c414c6d396ae8ccc0bfc9a1293b1602c/packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m#L215

Comment (hellohuanlin) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:46]: can you double check by setting a break point here and inspecting the type of `error`?

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 3 above)
Comment (hellohuanlin) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: hmmm, i thought it was ios 9 but i could remember it wrongly. Can you double check other plugins?

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 4 above)
Comment (hellohuanlin) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:226]: can you explain what is `stickyAuth` is? 

I believe the `result` must be called with a value or with an error, so that the dart side does not keep waiting (tho this is probably not related to your PR).

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 3 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: On the flutter docs it says minimum version is 11 https://docs.flutter.dev/deployment/ios#:~:text=The%20minimum%20iOS%20version%20that,to%20the%20highest%20required%20version.

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 5 above)
Comment (hellohuanlin) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:237]: > the original issue was happening because an error code was thrown that wasnt listed and it was failing silently. 

can you inspect the error code and explicitly list here? The `default` case shadows all potential future cases and may cause hard-to-find bugs.

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 4 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:226]: from what I can tell, stickyauth makes it so that when you leave then return to the app, the authentication screen persists. I didn't add it btw, I just moved the cases around

As for the result, that makes sense - should it return result(@No) instead? Since otherwise it would imply the authentication was successful

Comment (hellohuanlin) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:226]: I looked into the [doc](https://pub.dev/packages/local_auth#sticky-auth). It's intended behavior that we don't call the `result` callback here. So we can just leave it as it is.

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 2 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:46]: Screen Shot 2022-12-12 at 1 23 04 PM

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 3 above)
Comment (hellohuanlin) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: that's the current flutter version (people could be using older flutter)

Comment (hellohuanlin) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: quick search and found [this](https://github.com/flutter/flutter/issues/101965) - we are still at ios 9 here.

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 5 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:237]: the error code was 
https://github.com/flutter/plugins/blob/abb8e681c414c6d396ae8ccc0bfc9a1293b1602c/packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m#L232

So its already listed
Should I remove the default case then?

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 6]
```diff
@@ -41,6 +41,10 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
     [self deviceSupportsBiometrics:result];
   } else if ([@"isDeviceSupported" isEqualToString:call.method]) {
     result(@YES);
+  } else if ([@"handleAuthReplyWithSuccess" isEqualToString:call.method]) {
```
Comment (jmagman) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:44]: Hm, why was this added?  These should correspond to the messages coming from the dart API, like:
https://github.com/flutter/plugins/blob/2eb616545fff113141f9b0d40851a9322ad54c23/packages/local_auth/local_auth_platform_interface/lib/default_method_channel_platform.dart#L78-L79

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 5 above)
Comment (jmagman) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:237]: I agree, we should handle all enums and remove the `default` to detect (via compilation warning) when this problem happens again the next time we update the SDK.

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 3 above)
Comment (jmagman) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: Huan is right, this version of the plugin should continue to work on iOS 9 until the minimum version is bumped to 11.  The tests are passing because PRs only run on Flutter master, which is auto-migrating to iOS 11 so there's no issue.  I think that would even happen in stable too, which only gets run on post-submit.

Trying to think how to force the app to run against the iOS 9 SDK in CI.  I guess we just haven't run into issues related to this yet... cc @stuartmorgan

**Comment by jmagman** [comment]
Address the format issues:
https://cirrus-ci.com/task/6643342537129984?logs=format#L19

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 6 above)
Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m]: Oh I added it because  thats how the other tests were testing methods - let me figure out a way to test it without adding that bit then

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m (hunk 3 above)
Comment (stuartmorgan) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:223]: > Trying to think how to force the app to run against the iOS 9 SDK in CI.

Yeah, that's a problem given the auto-migrate logic. I would suggest someone just do a mass-update of all the iOS implementation packages to Flutter 3.3 and iOS 11 now (or maybe 3.3 first, and then iOS 11 as a second pass if that's non-trivial due to dead code removal).

That's the approach I've adopted for our old-Flutter-version support: when we update the N-2 stable version in CI, I mass-update all plugins to at least that minimum version, because we're no longer getting any automated testing that we aren't accidentally breaking them. (That kind of update doesn't cause any churn for users, because people still using old Flutter versions will just not be given the updates they can't use due to the resolver constraint on the Flutter version.)

File: packages/local_auth/local_auth/pubspec.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ description: Flutter plugin for Android and iOS devices to allow local
   authentication via fingerprint, touch ID, face ID, passcode, pin, or pattern.
 repository: https://github.com/flutter/plugins/tree/main/packages/local_auth/local_auth
 issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+label%3A%22p%3A+local_auth%22
-version: 2.1.2
+version: 2.1.3
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/pubspec.yaml:6]: No updates to `local_auth` are needed in this case, just the implementation package.

File: packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m [hunk 8]
```diff
@@ -27,451 +27,543 @@ @interface FLTLocalAuthPluginTests : XCTestCase
 @implementation FLTLocalAuthPluginTests
 
 - (void)setUp {
-  self.continueAfterFailure = NO;
+    self.continueAfterFailure = NO;
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:30]: You'll probably want to set your Xcode settings (or whatever editor you are using) to 2-space indent for ObjC :)

**Comment by stuartmorgan** [comment]
FYI if you install clang-format (go/clang-format-mac has pointers internally) you can just [run the format step locally](https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code) to get the format that CI is expecting.

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 9]
```diff
@@ -245,12 +247,7 @@ - (void)handleErrors:(NSError *)authError
   NSString *errorCode = @"NotAvailable";
   switch (authError.code) {
     case LAErrorPasscodeNotSet:
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"
-      // TODO(stuartmorgan): Remove the pragma and s/TouchID/Biometry/ in this constant when
-      // iOS 10 support is dropped. The values are the same, only the names have changed.
-    case LAErrorTouchIDNotEnrolled:
```
Comment (jmagman) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:252]: Does this need to be put back too then?

Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:252]: The macro or the comment?

Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:252]: Yes, missed that thanks!

File: packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m [hunk 10]
```diff
@@ -136,17 +136,112 @@ - (void)testFailedAuthWithBiometrics {
                                                                 @"localizedReason" : reason,
                                                               }];
 
+  XCTestExpectation *expectation = [self expectationWithDescription:@"Result is called"];
+  [plugin handleMethodCall:call
+                    result:^(id _Nullable result) {
+                      XCTAssertTrue([NSThread isMainThread]);
+                      XCTAssertTrue([result isMemberOfClass:[FlutterError class]]);
```
[the remark this answers is not in the document]
Comment (jmagman) [reply packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:143]: Bump this.

File: packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m [hunk 11]
```diff
@@ -216,26 +216,28 @@ - (void)handleAuthReplyWithSuccess:(BOOL)success
     result(@YES);
   } else {
     switch (error.code) {
-      case LAErrorPasscodeNotSet:
+      case LAErrorSystemCancel:
```
Comment (jmagman) [inline packages/local_auth/local_auth_ios/ios/Classes/FLTLocalAuthPlugin.m:219]: You could put this back to where it was at the bottom to avoid dirtying the git history and just add the `result(@NO)` part.

File: packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m [hunk 12]
```diff
@@ -179,8 +256,7 @@ - (void)testFailedAuthWithoutBiometrics {
   [plugin handleMethodCall:call
                     result:^(id _Nullable result) {
                       XCTAssertTrue([NSThread isMainThread]);
-                      XCTAssertTrue([result isKindOfClass:[NSNumber class]]);
-                      XCTAssertFalse([result boolValue]);
```
Comment (jmagman) [inline packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:183]: Did you mean to change these result blocks?

File: packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m [hunk 13]
```diff
@@ -220,10 +296,9 @@ - (void)testLocalizedFallbackTitle {
   XCTestExpectation *expectation = [self expectationWithDescription:@"Result is called"];
   [plugin handleMethodCall:call
                     result:^(id _Nullable result) {
-                      XCTAssertTrue([NSThread isMainThread]);
-                      XCTAssertTrue([result isKindOfClass:[NSNumber class]]);
                       OCMVerify([mockAuthContext setLocalizedFallbackTitle:localizedFallbackTitle]);
-                      XCTAssertFalse([result boolValue]);
```
Comment (jmagman) [inline packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:226]: Same.

File: packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m [hunk 14]
```diff
@@ -260,10 +335,9 @@ - (void)testSkippedLocalizedFallbackTitle {
   XCTestExpectation *expectation = [self expectationWithDescription:@"Result is called"];
   [plugin handleMethodCall:call
                     result:^(id _Nullable result) {
-                      XCTAssertTrue([NSThread isMainThread]);
-                      XCTAssertTrue([result isKindOfClass:[NSNumber class]]);
                       OCMVerify([mockAuthContext setLocalizedFallbackTitle:nil]);
-                      XCTAssertFalse([result boolValue]);
```
Comment (jmagman) [inline packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:266]: Same

Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:266]: So I changed it because in these tests, they were originally using a random errorCode of 99 which is not an enum value, so it was falling through the switch case and returning result(No) instead of throwing an error. I changed them to use appropriate error codes instead cuz I thought it would be more accurate to what the test is testing, but what do you think? Should I just leave it as it was before?

Comment (stuartmorgan) [reply packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:266]: For the second and third copy of this, the fact that it's simulating an error and asserting things about the error is a copypasta issue in these tests that I missed in review; there's no reason for a test about localized fallback title handling to be using the error path, or asserting anything other than the title part and the main thread. They were just created by copying and pasting from the closest test in the file it looks like. We should just update the dummy reply to a success path, and remove the irrelevant assertions.

For the first, if we intentionally have a codepath that returns NO instead of throwing an error, we should ideally test both that and the error path. The random error code test simulates what would happen if Apple added a new error that we didn't handle yet, for instance.

Comment (jmagman) [reply packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:266]: > The random error code test simulates what would happen if Apple added a new error that we didn't handle yet, for instance.

+1, there should be a test for a new enum that we haven't handled yet.

File: packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m [hunk 15]
```diff
@@ -175,6 +174,83 @@ - (void)testFailedAuthWithoutBiometrics {
...
+  [plugin handleMethodCall:call
+                    result:^(id _Nullable result) {
+                      XCTAssertTrue([NSThread isMainThread]);
+                      XCTAssertTrue([result isKindOfClass:[FlutterError class]]);
+                      [expectation fulfill];
+                    }];
+  [self waitForExpectationsWithTimeout:kTimeout handler:nil];
+}
+
+- (void)testFailedWithKnownErrorCode {
```
Comment (jmagman) [inline packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:187]: I think `testFailedAuthWithBiometrics` already handles a "known error code", this is more testing `LAErrorTouchIDNotEnrolled` explicitly, right?
```suggestion
- (void)testFailedTouchIDNotEnrolled {
```

Comment (LouiseHsu) [reply packages/local_auth/local_auth_ios/example/ios/RunnerTests/FLTLocalAuthPluginTests.m:187]: Youre right, it's kind of a dupe - Ill remove the test

**Comment by hellohuanlin** [comment]
I saw multiple email updates. Let me know when it's ready for a final review.

**Review by jmagman — APPROVED** [review]
LGTM!
1. You should update the description with `Fixes https://github.com/flutter/flutter/issues/103380` so GitHub will auto-link the issue to the PR and close the issue for you.  I manually linked them this time.
2. When you're ready to merge, put the `autosubmit` label on the issue and the bot will do the rest, and the plugin will be published.
